src/api/mod.rs

```diff
@@ -66,101 +66,191 @@ pub trait AsyncElement {
     fn process(&mut self, packet: Self::Input) -> Self::Output;
 }
 
-pub struct AsyncElementFrontend<E: AsyncElement> {
-    input_stream: ElementStream<E::Input>,
-    channel: Sender<E::Output>,
-    element: E
+/// The AsyncElementLink is a wrapper to create and contain both sides of the
+/// link, the consumer, which intakes and processes packets, and the provider,
+/// which provides an interface where the next element retrieves the output
+/// packet.
+pub struct AsyncElementLink< E: AsyncElement> {
+    pub consumer: AsyncElementConsumer<E>,
+    pub provider: AsyncElementProvider<E>
 }
 
-pub struct AsyncElementBackend<E: AsyncElement> {
-    channel: Receiver<E::Output>
-}
+impl<E: AsyncElement> AsyncElementLink<E> {
+    pub fn new(input_stream: ElementStream<E::Input>, element: E, queue_capacity: usize) -> Self {
 
-pub struct AsyncElementLink< E: AsyncElement> {
-    pub frontend: AsyncElementFrontend<E>,
-    pub backend: AsyncElementBackend<E>
-}
+        let (to_provider, from_consumer) = bounded::<Option<E::Output>>(queue_capacity);
```

Comment:
This is not a blocker, but I think `bounded` here is a difficult to read, and would be much clearer as `crossbeam_channel::bounded`

Reply:
That is a good idea, it is a good idea to keep calls out to dependent libraries explicit, for clarity. Especially since bounded is already quite non-descriptive